universal-hash/src/lib.rs

```diff
@@ -18,167 +18,153 @@
 //! [Universal Hash Functions]: https://en.wikipedia.org/wiki/Universal_hashing
 
 #![no_std]
-#![forbid(unsafe_code)]
 #![doc(
     html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
     html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/8f1a9894/logo.svg",
-    html_root_url = "https://docs.rs/universal-hash/0.4.1"
+    html_root_url = "https://docs.rs/universal-hash/0.5.0"
 )]
+#![cfg_attr(docsrs, feature(doc_cfg))]
 #![warn(missing_docs, rust_2018_idioms)]
 
 #[cfg(feature = "std")]
 extern crate std;
 
-pub use generic_array::{self, typenum::consts};
+pub use crypto_common::{
+    self, generic_array,
+    typenum::{self, consts},
+    Block, Key, KeyInit, ParBlocks,
+};
 
-use generic_array::typenum::Unsigned;
+use core::slice;
+use crypto_common::{BlockSizeUser, ParBlocksSizeUser};
 use generic_array::{ArrayLength, GenericArray};
-use subtle::{Choice, ConstantTimeEq};
-
-/// Keys to a [`UniversalHash`].
-pub type Key<U> = GenericArray<u8, <U as NewUniversalHash>::KeySize>;
-
-/// Blocks are inputs to a [`UniversalHash`].
-pub type Block<U> = GenericArray<u8, <U as UniversalHash>::BlockSize>;
-
-/// Instantiate a [`UniversalHash`] algorithm.
-pub trait NewUniversalHash: Sized {
-    /// Size of the key for the universal hash function.
-    type KeySize: ArrayLength<u8>;
+use subtle::ConstantTimeEq;
+use typenum::Unsigned;
+
+/// Trait implemented by UHF backends.
+pub trait UhfBackend: ParBlocksSizeUser {
+    /// Process single block.
+    fn proc_block(&mut self, block: &Block<Self>);
+
+    /// Process several blocks in parallel.
+    #[inline(always)]
+    fn proc_par_blocks(&mut self, blocks: &ParBlocks<Self>) {
+        for block in blocks {
+            self.proc_block(block);
+        }
+    }
+}
 
-    /// Instantiate a universal hash function with the given key.
-    fn new(key: &Key<Self>) -> Self;
+/// Trait for [`UhfBackend`] users.
+///
+/// This trait is used to define rank-2 closures.
+pub trait UhfClosure: BlockSizeUser {
+    /// Execute closure with the provided UHF backend.
+    fn call<B: UhfBackend<BlockSize = Self::BlockSize>>(self, backend: &mut B);
 }
 
 /// The [`UniversalHash`] trait defines a generic interface for universal hash
 /// functions.
-pub trait UniversalHash: Clone {
-    /// Size of the inputs to and outputs from the universal hash function
-    type BlockSize: ArrayLength<u8>;
+pub trait UniversalHash: BlockSizeUser + Sized {
+    /// Update hash function state using the provided rank-2 closure.
```

Comment:
Document how this method is intended to be implemented (calling out to target-specific backends). If you don't want that documentation to live on the trait that end users will be using, then instead add a reference here to `UhfBackend` and document the implementor notes there.

Reply:
This is also where the note about user trait usage should go, for e.g. AEADs (https://github.com/RustCrypto/AEADs/pull/414#issuecomment-1107614537).